test/hotspot/jtreg/runtime/cds/appcds/aotCache/AOTCacheSupportForCustomLoaders.java

```diff
@@ -80,8 +106,41 @@ public static void main(String args[]) throws Exception {
         } catch (VerifyError ve) {
             System.out.println("Expected: " + ve);
         }
+    }
 
-        // TODO: more test cases JDK-8354557
+    // Test 3: custom loader defines a class from the exact location as a class defined in the boot layer.
+    static void test3(String modulePath) throws Exception {
+        Class<?> c0 = Class.forName("com.test.Foo");
+        System.out.println(c0);
+        System.out.println(System.identityHashCode(c0.getModule()));
+        System.out.println(c0.getModule().getName());
+        System.out.println(c0.getClassLoader());
+
+        // Regression test for JDK-8356838
+        //
+        // We create a new layer that loads the com.test module from the modulePath into
+        // a different class loader.
+        ModuleFinder finder = ModuleFinder.of(Paths.get(modulePath));
+        ModuleLayer parent = ModuleLayer.boot();
+        Configuration cf = parent.configuration().resolve(finder, ModuleFinder.of(), Set.of("com.test"));
+        ClassLoader scl = ClassLoader.getSystemClassLoader();
+        ModuleLayer layer = parent.defineModulesWithOneLoader(cf, scl);
+        Class<?> c1 = layer.findLoader("com.test").loadClass("com.test.Foo");
+
+        System.out.println(c1);
+        System.out.println(System.identityHashCode(c1.getModule()));
+        System.out.println(c1.getModule().getName());
+        System.out.println(c1.getClassLoader());
+
+        if (!c1.getModule().getName().equals("com.test")) {
+            throw new RuntimeException("Unexpected module: " + c1.getModule());
+        }
+        if (c1.getModule() == c0.getModule()) {
+            throw new RuntimeException("Unexpected class loader: " + c1.getClassLoader());
```

Comment:
Should the `throw` statement be the following?

`throw new RuntimeException("Unexpected module: " + c1.getModule());`